src/lib/util/textfile.c

```diff
@@ -190,3 +191,101 @@ textfile_write(const char *filepath,
 
     return ret;
 }
+
+errno_t
+textfile_copy(const char *source,
+              const char *destdir,
+              const char *destname,
+              mode_t dir_mode)
+{
+    struct stat statbuf;
+    FILE *fsource = NULL;
+    FILE *fdest = NULL;
+    size_t bytes_written;
+    size_t bytes_read;
+    mode_t oldmask;
+    char *destpath;
+    char buf[32];
+    errno_t ret;
+
+    ret = stat(source, &statbuf);
+    if (ret == -1) {
+        ret = errno;
+        return ret;
+    }
+
+    ret = file_make_path(destdir, dir_mode);
+    if (ret != EOK) {
+        return ret;
+    }
+
+    destpath = format("%s/%s", destdir, destname);
+    if (destpath == NULL) {
+        return ENOMEM;
+    }
+
+    /* Temporary umask before we change the owner and permissions. */
+    oldmask = umask(0600);
+
+    fsource = fopen(source, "r");
```

Comment:
First calling stat and then doing open is not the way files should be operated with. You should first just open the file and then read the fd from the handle using fileno and then instead of calling stat on the path, call fstat on the fd. This way, an attacker cannot change the file you are referencing by path.

(Yes, it is unlikely someone will do this, but let's use best practices everywhere)

Reply:
Thank you for the hint.